scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -39,119 +41,87 @@ class ScaleOutRelayService(
+      _ <- woLDAPNodeGroupRepository.create(nodeGroup, categoryId, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueId, directDistribPolicy, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueIdCommon, commonDirective, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleSetup, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleDistribPolicy, modId, actor, reason)
     } yield {
       nodeInf
     }
+
+    promote.catchAll { err =>
+      PluginLogger.error(err)
+      IO.effect(rollbackPromote(ruleTarget, ruleSetup, ruleDistribPolicy, nodeInf, nodeGroup, activeTechniqueId, activeTechniqueIdCommon, modId, actor))
+    }
   }
 
-//  private[scaleoutrelay] def rollbackFromLevel(level : Int, err : RudderError,ruleTarget:  PolicyServerTarget,ruleSetup : Rule,ruleDistribPolicy: Rule,nodeInf: NodeInfo, nodeGroup: NodeGroup, activeTechniqueId: ActiveTechniqueId, activeTechniqueIdCommon: ActiveTechniqueId,modId: ModificationId, actor: EventActor,) = {
-//    val e = Chained(s"Error when promoting '${nodeInf.node.id.value}' to relay, trying to restore initial node configuration", err)
-//    PluginLogger.error(e)
-//    for {
-//      _ <- if(level >= 1) woLDAPNodeRepository.deleteNode(nodeInf.node, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed ${nodeInf.node} failed"))
-//      _ <- if(level >= 2) woLDAPNodeGroupRepository.deletePolicyServerTarget(ruleTarget).chainError(s"Rollback promote to relay failed : removed rule target ${ruleTarget.target} failed")
-//      _ <- if(level >= 3) woLDAPNodeGroupRepository.delete(nodeGroup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed node group ${nodeGroup.id.value} failed")
-//      _ <- if(level >= 4) woDirectiveRepository.deleteActiveTechnique(activeTechniqueId, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed active technique ${activeTechniqueId.value} failed")
-//      _ <- if(level >= 5) woDirectiveRepository.deleteActiveTechnique(activeTechniqueIdCommon, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed active technique ${activeTechniqueIdCommon.value} failed")
-//      _ <- if(level >= 6) woRuleRepository.delete(ruleSetup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleSetup.id.value} failed")
-//      _ <- if(level >= 7) woRuleRepository.delete(ruleDistribPolicy.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleDistribPolicy.id.value} failed")
-//    } yield ()
-//  }
+  private[scaleoutrelay] def rollbackPromote(
+      ruleTarget              : PolicyServerTarget
```

Comment:
details here but you only need nodeInfo, modId and actor

you only need ids which can be recreated from your nodeId for the other data

Reply:
Still need to be updated, you only need the nodeInfo, even the initial node info is needed